File: data structures and algorithms/java/08_059.java
import java.util.*;

public class KeyCollectionSequence {
   public static int minimumStepsToCollectKeys(
         String[] maze) {
      int rows = maze.length;
      int cols = maze[0].length();
      
      // Find start and keys
      int startRow = 0, startCol = 0;
      List<Character> keys = new ArrayList<>();
      
      for (int r = 0; r < rows; r++) {
         for (int c = 0; c < cols; c++) {
            char cell = maze[r].charAt(c);
            if (cell == 'S') {
               startRow = r;
               startCol = c;
            } else if ('a' <= cell && cell <= 'z') {
               keys.add(cell);
            }
         }
      }
      
      // Sort keys alphabetically
      Collections.sort(keys);
      
      // Movement directions
      int[][] directions = {{-1, 0}, {0, 1}, 
                             {1, 0}, {0, -1}};
      
      // Track current position
      int currRow = startRow, currCol = startCol;
      int totalSteps = 0;
      Set<Character> collectedKeys = new HashSet<>();
      
      // Collect keys in order
      for (char key : keys) {
         // BFS to find key path
         Queue<int[]> queue = new LinkedList<>();
         queue.add(new int[]{currRow, currCol, 0});
         Set<String> visited = new HashSet<>();
         visited.add(currRow + "," + currCol);
         boolean found = false;
         
         while (!queue.isEmpty() && !found) {
            int[] curr = queue.poll();
            int r = curr[0], c = curr[1], steps = curr[2];
            
            // Key found
            if (maze[r].charAt(c) == key) {
               totalSteps += steps;
               currRow = r;
               currCol = c;
               collectedKeys.add(key);
               found = true;
               break;
            }
            
            // Try all directions
            for (int[] dir : directions) {
               int nr = r + dir[0], nc = c + dir[1];
               
               // Validate new position
               if (0 <= nr && nr < rows && 
                   0 <= nc && nc < cols && 
                   maze[nr].charAt(nc) != '#' && 
                   !visited.contains(nr + "," + nc)) {
                  
                  // Check door keys
                  char cell = maze[nr].charAt(nc);
                  if ('A' <= cell && cell <= 'Z') {
                     char keyNeeded = 
                        Character.toLowerCase(cell);
                     if (!collectedKeys.contains(
                           keyNeeded)) {
                        continue;
                     }
                  }
                  
                  queue.add(new int[]{nr, nc, 
                                      steps + 1});
                  visited.add(nr + "," + nc);
               }
            }
         }
         
         // Key unreachable
         if (!found) {
            return -1;
         }
      }
      
      return totalSteps;
   }
}

File: data structures and algorithms/cs/08_059.cs
using System;
using System.Collections.Generic;

public class KeyCollectionSequence
{
   public static int MinimumStepsToCollectKeys(
      string[] maze)
   {
      int rows = maze.Length;
      int cols = maze[0].Length;

      int startRow = 0, startCol = 0;
      var keys = new List<char>();

      for (int r = 0; r < rows; r++)
      {
         for (int c = 0; c < cols; c++)
         {
            char cell = maze[r][c];
            if (cell == 'S')
            {
               startRow = r;
               startCol = c;
            }
            else if ('a' <= cell && cell <= 'z')
            {
               keys.Add(cell);
            }
         }
      }

      keys.Sort();

      int[][] directions =
      {
         new[] { -1, 0 },
         new[] { 0, 1 },
         new[] { 1, 0 },
         new[] { 0, -1 }
      };

      int currRow = startRow, currCol = startCol;
      int totalSteps = 0;
      var collectedKeys = new HashSet<char>();

      foreach (var key in keys)
      {
         var queue = new Queue<int[]>();
         queue.Enqueue(new[] { currRow, currCol, 0 });
         var visited = new HashSet<string>
         {
            currRow + "," + currCol
         };
         bool found = false;

         while (queue.Count > 0 && !found)
         {
            var curr = queue.Dequeue();
            int r = curr[0], c = curr[1], steps = curr[2];

            if (maze[r][c] == key)
            {
               totalSteps += steps;
               currRow = r;
               currCol = c;
               collectedKeys.Add(key);
               found = true;
               break;
            }

            foreach (var dir in directions)
            {
               int nr = r + dir[0], nc = c + dir[1];

               if (0 <= nr && nr < rows &&
                  0 <= nc && nc < cols &&
                  maze[nr][nc] != '#' &&
                  !visited.Contains(nr + "," + nc))
               {
                  char cell = maze[nr][nc];
                  if ('A' <= cell && cell <= 'Z')
                  {
                     char keyNeeded = char.ToLower(cell);
                     if (!collectedKeys.Contains(keyNeeded))
                     {
                        continue;
                     }
                  }

                  queue.Enqueue(new[] { nr, nc, steps + 1 });
                  visited.Add(nr + "," + nc);
               }
            }
         }

         if (!found)
         {
            return -1;
         }
      }

      return totalSteps;
   }
}